### PR TITLE
Fixed the Filter issue on the catalog page

### LINKED ISSUE #1806: Catalog filter not working as expected
### Description
On clicking technology it filters the data however on unclicking it it should filter again which it is not doing, also if I select technology I can select `Categories` and if I selected `Categories` then it will get deselect technology filters.

As you can see in video, it shows number `3` after Thanos technology however on click it only renders one design, either number is wrong or rendering

https://github.com/meshery/meshery.io/assets/71271069/c190de0b-06cc-4f5d-addd-384625ea31ce

### Expected Behavior
- [ ] User should be allowed to select both categories and technology
- [ ] On deselect trigger filter again to update rendered designs

### Enviroment:
 - OS: Mac Linux Windows
 - Browser: Chrome Safari Firefox
 - Version: 
 - Device: Desktop Mobile

---
## Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)

The meshery.io website uses Jekyll and GitHub Pages. Site content is found under the [`master` branch](https://github.com/meshery/meshery.io/tree/master).
- 📚 See the [Contributing to Meshery.io Website](https://github.com/meshery/meshery.io#contributing-to-the-mesheryio-website) section of the readme.md and other [contributing instructions](https://docs.meshery.io/project/contributing), too.
- 🎨 Wireframes and [designs for Meshery UI](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI) in Figma [(open invite)](https://www.figma.com/team_invite/redeem/qJy1c95qirjgWQODApilR9)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](https://discuss.layer5.io) and [Community Slack](https://slack.meshery.io)

### PR DESCRIPTION
**Pull request by kanishk6103**
**Description**

This PR fixes #1806

**Notes for Reviewers**

**[Signed commits](https://docs.meshery.io/project/contributing#signing-off-on-commits-developer-certificate-of-origin)**
- [x] User should be allowed to select both categories and technology
- [x] On deselect trigger filter again to update rendered designs
- [x] The count of each technology are now correctly displayed
- [x] Yes, I signed my commits.

### The PR fixes the following issues: 
- Users can now filter by technologies and upon unchecking all technologies, the filter works correctly.
- Users can select technologies and categories at the same time and the filters are applied according to both.
- If no technology is selected then the filter is applied according to the categories alone.
- The Technologies count is now fixed and the correct numbers are now displayed.

A video demonstration for the same:

https://github.com/meshery/meshery.io/assets/72643990/a34b43bb-a70e-4197-aa99-257a0b046f63

### CODE REVIEW

**Comment by Deepak0320Singhal** [comment]
@kanishk6103 Thank you for your contribution.
Let's discuss this on Websites' call. Adding this as an agenda item into the [meeting minutes](https://docs.google.com/document/d/1XczAHXVe2FIWPqiF57ospJ43zw5cZQ7ui8mn39v5EvA/edit).

**Comment by kanishk6103** [comment]
> @kanishk6103 Thank you for your contribution. Let's discuss this on Websites' call. Adding this as an agenda item into the [meeting minutes](https://docs.google.com/document/d/1XczAHXVe2FIWPqiF57ospJ43zw5cZQ7ui8mn39v5EvA/edit).

can you please review the changes? @Deepak0320Singhal

File: _includes/patterns-filter.html [hunk 1]
```diff
@@ -253,13 +253,19 @@
...
   const categoryContainer = document.getElementById("category-div");
-  
+
+  // globalTechnologyList stores all the technologies checked by user as strings
+  // example: ["gcp", "keda"]
+  let globalTechnologyList = [];
+
+  // globalCategorySelected stores the category selected by the user since only one can be selected at a time we don't need any array
+  // initially it is set to "All"
+  let globalCategorySelected = "All";
```
Comment (Yashsharma1911) [inline _includes/patterns-filter.html:263]: In similar wane we can support multiple category selection, can you ensure that we do that too

File: _includes/patterns-filter.html [hunk 2]
```diff
@@ -403,32 +413,41 @@
...
+            const filter = patternCards[j].getAttribute("type").toLowerCase();
+            let showCard = false;
+            if (filter.includes(value.toLowerCase())) {
+              showCard = true;
+            }
+            if (showCard) {
+              patternCards[j].style.display = "block";
+            } else {
+              patternCards[j].style.display = "none";
+            }
```
Comment (Yashsharma1911) [inline _includes/patterns-filter.html:439]: This if else condition seems to be using multiple time, maybe we can extract it to a function

File: _includes/patterns-filter.html [hunk 3]
```diff
@@ -347,52 +354,55 @@
...
     } else {
 
-      // Loop over all pattern cards and check if any of the filter in filterArray is present in pattern card's filter attribute
-      // If yes, the show the card, else hide it
+      // Loop over all pattern cards and check if any of the filter in globalTechnologyList is present in pattern card's filter attribute
+      // If yes then set showCardTech to true
+      // also check if the selected category matches the card type (category the card belongs to)
+      // If yes then set showCardCat to true
+    
+      // If both showCardTech and showCardCat are true, then show the card otherwise hide it
```
Comment (Yashsharma1911) [inline _includes/patterns-filter.html:380]: Thanks for making sure comments are updated

**Review by Yashsharma1911 — CHANGES_REQUESTED** [review]
Thankyou @kanishk6103 for making sure it get's fixes quickly, I have pointed out few things after we should be good to go and should mere your PR

**Comment by Yashsharma1911** [comment]
Thanks for adding checklist in PR it is always helpful

File: _includes/patterns-filter.html (hunk 1 above)
Comment (kanishk6103) [reply _includes/patterns-filter.html:263]: I can implement that feature, for that I will have to convert those radio buttons to checkboxes, shall I go ahead and do that?

File: _includes/patterns-filter.html (hunk 3 above)
Comment (kanishk6103) [reply _includes/patterns-filter.html:380]: Thank you :D

File: _includes/patterns-filter.html (hunk 1 above)
Comment (kanishk6103) [reply _includes/patterns-filter.html]: Hey I have updated the branch, we can now filter while selecting multiple categories! @Yashsharma1911

**Comment by kanishk6103** [comment]
> Thanks for adding checklist in PR it is always helpful

Hey, can you please check it now? I have added multiple category selection @Yashsharma1911

File: _includes/patterns-filter.html (hunk 1 above)
Comment (Yashsharma1911) [reply _includes/patterns-filter.html]: Thankyou Kanishk

**Comment by Yashsharma1911** [comment]
@kanishk6103  can you remove `all` from here as now it is a checkbox, by default show designs with "All", if no checkbox is selected mean "All", similar to meshery.layer5.io/catalog

Screenshot 2024-07-11 at 5 57 11 PM

**Comment by Yashsharma1911** [comment]
We can also show numbers ahead of `Categories` similar to technologies something maybe you can address in next PR till we will merge this

**Comment by Yashsharma1911** [comment]
Other things looks good to me, ping me again when you will make requested changes ^^

**Comment by kanishk6103** [comment]
> We can also show numbers ahead of `Categories` similar to technologies something maybe you can address in next PR till we will merge this

Sure! For now I will have to just remove the "All" category button right? @Yashsharma1911

**Comment by kanishk6103** [comment]
Hey @Yashsharma1911! I have made the required changes, please review them.

**Review by Yashsharma1911 — APPROVED** [review]
Thanks LGTM now

**Comment by Yashsharma1911** [comment]
@kanishk6103 as you enhance this more in next PR please consider to have an empty state here too. ^^

Screenshot 2024-07-12 at 1 34 25 AM

**Comment by Yashsharma1911** [comment]
Yay 🥳 congrats on you first PR merge

**Comment by kanishk6103** [comment]
> @kanishk6103 as you enhance this more in next PR please consider to have an empty state here too. ^^

Sure!

**Comment by kanishk6103** [comment]
> Yay 🥳 congrats on you first PR merge

Thank you :D   
I hope to contribute more this community

**Pull request merged**